### ISSUE TITLE
support unicode character for JavaTokenMarker

### ISSUE THREAD

**Issue by Yun-Shan** [bobbylight/RSyntaxTextArea#355]
I'm sorry that my English is very poor and I use machine translation

now: JavaTokenMarker Identifier not support unicode character, only support \uxxxx, a valid unicode character identifier will be defined as an error_identifier

expect: JavaTokenMarker Identifier should support all valid identifier, include unicode character.

**current code:**
https://github.com/bobbylight/RSyntaxTextArea/blob/f03e87801156d406a2e922a831213e36c4314396/RSyntaxTextArea/src/main/java/org/fife/ui/rsyntaxtextarea/modes/JavaTokenMaker.flex#L238-L239
https://github.com/bobbylight/RSyntaxTextArea/blob/f03e87801156d406a2e922a831213e36c4314396/RSyntaxTextArea/src/main/java/org/fife/ui/rsyntaxtextarea/modes/JavaTokenMaker.flex#L214-L215

---
**try fix:**
The following is the code I tried to fix this problem, but it can’t run on JFlex 1.4.1. I use JFlex 1.8.2 to generate the code and refer to the [wiki](https://github.com/bobbylight/RSyntaxTextArea/wiki/Adding-Syntax-Highlighting-for-a-new-Language) to delete the two methods defined repeatedly. But I’m not sure if there is a problem I didn’t notice

src/main/java/org/fife/ui/rsyntaxtextarea/modes/JavaTokenMaker.flex#L238-L239
```flex
IdentifierStart					= (\p{L}|\p{Letter_Number}|\p{Currency_Symbol}|\p{Connector_Punctuation})
IdentifierPart					= ({IdentifierStart}|{Digit}|\p{Combining_Mark}|\p{Non_Spacing_Mark}|[\u0000-\u0008]|[\u000E-\u001B]|[\u007F-\u009F]|\p{Format}|("\\"{EscapedSourceCharacter}))
```
src/main/java/org/fife/ui/rsyntaxtextarea/modes/JavaTokenMaker.flex#L214
(it will throw an ArrayIndexOutOfBoundsException if `s.count - 1`)
(JFlex 1.8.2 don't generate a `zzPushbackPos` field)
```java
zzEndRead = zzStartRead + s.count;
zzCurrentPos = zzMarkedPos = s.offset;
```
---
**refer these javadoc**
java.lang.Character#isJavaIdentifierStart(int)
```javadoc
Determines if the character (Unicode code point) is permissible as the first character in a Java identifier.
A character may start a Java identifier if and only if one of the following conditions is true:
- isLetter(codePoint) returns true
- getType(codePoint) returns LETTER_NUMBER
- the referenced character is a currency symbol (such as '$')
- the referenced character is a connecting punctuation character (such as '_').
```
java.lang.Character#isJavaIdentifierPart(int)
```javadoc
Determines if the character (Unicode code point) may be part of a Java identifier as other than the first character.
A character may be part of a Java identifier if any of the following are true:
- it is a letter
- it is a currency symbol (such as '$')
- it is a connecting punctuation character (such as '_')
- it is a digit
- it is a numeric letter (such as a Roman numeral character)
- it is a combining mark
- it is a non-spacing mark
- isIdentifierIgnorable(codePoint) returns true for the character
```
---
**refer wikipedia for Regular_expression**
Chinese Version has an unicode character properties list, but i can't find a similar list at English Version
[English Version](https://en.wikipedia.org/wiki/Regular_expression#Unicode)
[Chinese Version](https://zh.wikipedia.org/wiki/%E6%AD%A3%E5%88%99%E8%A1%A8%E8%BE%BE%E5%BC%8F#Unicode%E5%A4%84%E7%90%86)

**Comment by bobbylight** [comment]
This is correct, see for example [this discussion](https://stackoverflow.com/questions/32417922/java-valid-identifier-from-java-language-specification) from a quick Google search.

**Comment by bobbylight** [comment]
As for the JFlex version being pinned to 1.4.1 - yeah it sucks, but the code it generates changed in the 1.5.x release and I couldn't work out what manual changes were needed to get it working per-line like we have with the 1.4.x-generated code.

I've thought about bundling JFlex 1.4.1, or pulling it down as an unshipped dependency along with a gradle task to generate the TokenMaker source files.  It might be time to revisit that.

In the mean time I'll look into ways to referene these unicode properties.  It looks like I can use `[:jletter:]` and `[:jletterdigit:]`, which correspond directly to `Character.isJavaIdentifierStart()` and `Character.isJavaIdentifierPart()`.  I'll do some experimentation and see.

**Comment by bobbylight** [comment]
This fix is now in the latest 3.1.3-SNAPSHOT in Sonatype.  3.1.3 will probably be released to Maven Central after a couple other bugs are fixed.
